avro-java8/src/main/java/com/fasterxml/jackson/dataformat/avro/java8/AvroJavaTimeAnnotationIntrospector.java

```diff
@@ -0,0 +1,127 @@
+package com.fasterxml.jackson.dataformat.avro.java8;
+
+import com.fasterxml.jackson.core.Version;
+import com.fasterxml.jackson.databind.introspect.Annotated;
+import com.fasterxml.jackson.dataformat.avro.AvroAnnotationIntrospector;
+import com.fasterxml.jackson.dataformat.avro.AvroType;
+import com.fasterxml.jackson.dataformat.avro.java8.deser.LocalDateDeserializer;
+import com.fasterxml.jackson.dataformat.avro.java8.deser.LocalDateTimeDeserializer;
+import com.fasterxml.jackson.dataformat.avro.java8.deser.LocalTimeDeserializer;
+import com.fasterxml.jackson.dataformat.avro.java8.deser.OffsetDateTimeDeserializer;
+import com.fasterxml.jackson.dataformat.avro.java8.deser.ZonedDateTimeDeserializer;
+import com.fasterxml.jackson.dataformat.avro.java8.ser.LocalDateSerializer;
+import com.fasterxml.jackson.dataformat.avro.java8.ser.LocalDateTimeSerializer;
+import com.fasterxml.jackson.dataformat.avro.java8.ser.LocalTimeSerializer;
+import com.fasterxml.jackson.dataformat.avro.java8.ser.OffsetDateTimeSerializer;
+import com.fasterxml.jackson.dataformat.avro.java8.ser.ZonedDateTimeSerializer;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.OffsetDateTime;
+import java.time.ZonedDateTime;
+
+class AvroJavaTimeAnnotationIntrospector extends AvroAnnotationIntrospector {
+  static final AvroJavaTimeAnnotationIntrospector INSTANCE = new AvroJavaTimeAnnotationIntrospector();
+
+  @Override
+  public Object findSerializer(Annotated a) {
+    AvroType logicalType = _findAnnotation(a, AvroType.class);
+    if (null != logicalType) {
+      switch (logicalType.logicalType()) {
+        case TIMESTAMP_MILLISECOND:
```

Comment:
timestamps have no timezone, explicit or implicit. it is more correct to model them using java.time.Instant. this can be converted into any XXXDateTime class by providing a timezone, or stating it is implicit - this information is not available from the encoded timestamp itself

Reply:
@lukejackson That makes sense because AVRO stores everything without timezone as well. With jackson I was under the impression that I had to create a serializer and deserializer for each of the java.time.* classes. Is there another approach that I can take?

Reply:
My suggestion would be to add java.time.Instant support - this is arguably the canonical representation of timestamps in Java 8+.

As for the other XXXDateTime types, I think there is an argument for them not to be supported, given there is no corresponding type in Avro. 

If you do introduce a conversion, it will introduce information not on the wire (e.g. timezone) and also be lossy, given a date time with a time zone can map to two different timestamps during daylight savings time transitions. 

I would argue that given these limitations, it should be left to the user to perform the conversion and understand their assumptions and associated limitations.

Note the closest Avro does have is its date and time-xxx logical types. To model a LocalDateTime in Avro requires two fields with these corresponding types. To model a ZonedDateTIme requires a third field with the zone id.

Reply:
I get what you are saying now. Not having Instant support was an oversight. My last comment I thought you were telling me to convert to instant and delegate that to a serializer. I'll add instant support. LocalDateTime and LocalDate would be the closest thing to the avro specification given that neither Local* or the avro specification include timezone information.

Reply:
I can see how LocalDate is the closest to the avro spec for date, but not sure I understand what you mean by LocalDateTime - I don't think there is anything in the avro spec that relates to that?

Don't want to sound nit picking, and appreciate your effort here. The concepts behind the date and time types are deceptively complicated. Given recent stagnation in the Avro project (and there is a long standing PR to add java.time there), I am hoping that this Jackson module could be a workable alternative to core Avro for some of my use cases.

Reply:
This is what I was thinking when I started this PR. The Local* time classes go not have timezone information so they translate perfectly to the corresponding avro logical types. Not having instant was just an oversight. I've already started on adding it to this pull. Here is what I'm thinking. Thinking about your comments on OffsetDateTime and ZonedDateTime. You are correct. I was planning on doing some lossy conversions to support these types. Alternatively I could create logical types for them. The point of logical types in Avro is to support things outside of the specification. To do this I could create a logical type called `zoned-date`, `zoned-time-millis`, `zoned-time-micros`, `zoned-timestamp-millis`, and `zoned-timestamp-millis`. This would allow me to store the timezone information along with the date/timestamp information.  

This was my original thoughts.

| Java Type      | Avro Logical Type                  | Notes                                                             |
|----------------|------------------------------------|-------------------------------------------------------------------|
| LocalDate      | Date                               | Lossless conversion                                               |
| LocalTime      | time-millis, time-micros           | Lossless conversion                                               |
| LocalDateTime  | timestamp-millis, timestamp-millis | Lossless conversion                                               |
| Instant        | timestamp-millis, timestamp-millis | Lossless conversion                                               |
| OffsetDateTime | timestamp-millis, timestamp-millis | Lossy conversion? Custom logical type? Timestamp adjusted to UTC. |
| ZonedDateTime  | timestamp-millis, timestamp-millis | Lossy conversion? Custom logical type? Timestamp adjusted to UTC. |
| OffsetTime     | time-millis, time-micros           | Lossy conversion? Custom logical type? Timestamp adjusted to UTC. |

Reply:
I hadn't thought about adding new logical types to Avro.

I think the way forward depends on what you're looking to achieve. If it's a way to represent all the Java related time classes in Avro - i.e. the Java POJO is the canonical schema and the Avro representation is just derived from that - then I can understand you'd look for a way to encode them in Avro. There are some options:

1. reusing another type which is close in meaning (e.g. timestamp-millis for LocalDateTime and have any consumers - which may not be this library - know that they should convert the timestamp into a ZonedDateTime with UTC time zone and convert to LocalDateTime)
2. adding your own logical types. this would mean that any consumer - which again may not be this library - would also have to implement them.
3. using the record type composed of types supported by Avro. this is what Jackson does when converting POJOs to JSON - they become JSON dictionaries. Avro records would be the equivalent here. e.g. LocalDateTime would be a record containing a date field and a time-millis/time-micros field.

If instead you are looking for a way to directly represent Avro records as Java classes - i.e. the Avro schema is the canonical schema and the Java POJOs are just derived from that - then you just need to support the Avro types and no more, and leave it to the user to define their schemas to model the data they want to encode (e.g. manually define a record representing local date and local time), or let them choose how to shoehorn the value into an existing Avro type (e.g. represent as a string, as a long, etc.)

For my use cases, the Avro schemas are the canonical schemas. As such I only use the Avro types. This is because I use Avro to communicate between applications written in multiple languages, such as Java, C++, Python, and I rely on Kafka tools that only support Avro types (e.g. Connect and its various sinks).

Where the Avro types don't support my use case I either use a different type (e.g. long for nano precision timestamps) or multiple fields (e.g. combination of date and time-millis for a local date time), and I'm happy to accept that these fields would appear as such in Java, C++, Python, my database tables, etc.